### PR TITLE
Fix #6510 ?

### LINKED ISSUE #6510: Thread safety in TPad
## Describe the bug
Crash.
```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff59494f8 in TPad::ShowGuidelines (this=0x249da40, object=0x249da40, event=21, mode=114 'r', cling=true) at /home/andriish/Projects/ROOT/root/graf2d/gpad/src/TPad.cxx:6248
6248	      tmpGuideLinePad->Delete();
(gdb) bt
#0  0x00007ffff59494f8 in TPad::ShowGuidelines(TObject*, int, char, bool) (this=0x249da40, object=0x249da40, event=21, mode=114 'r', cling=true) at /Projects/ROOT/root/graf2d/gpad/src/TPad.cxx:6248
#1  0x00007ffff5932f5f in TPad::ExecuteEvent(int, int, int) (this=0x249da40, event=21, px=412, py=186) at  /Projects/ROOT/root/graf2d/gpad/src/TPad.cxx:2159

```

## Expected behavior
No crash.

## To Reproduce
No working standalone reproducer so far.

## Setup
master ROOT with gcc4.8.5 on CentOS7

## Additional context
The relevant code is 

```
   static TPad * tmpGuideLinePad;

   //delete all existing Guidelines and create new invisible pad
   if (tmpGuideLinePad) {
      if (object == tmpGuideLinePad) { // in case of funny button click combination.
         tmpGuideLinePad->Delete();
         tmpGuideLinePad = 0;
         return;
      }
      tmpGuideLinePad->Delete();
      tmpGuideLinePad = 0;
   }

```
Obviously the code if completely ok for a single thread application.
However, a protection for multithread apps is needed when dealing with tmpGuideLinePad.

Tag @couet

### PR DESCRIPTION
**Pull request by andriish**
Hopefully this will fix #6510 

Tag @couet

### CODE REVIEW

**Comment by couet** [comment]
"Hopefully" ? you cannot test ? I must admit I am not sure to understand this  fix.

**Comment by andriish** [comment]
Hi @couet ,

I've tested it and it fixes things for me.
I wrote hopefully because I cannot write a simple reproducer.    
Please read it as "fixes for my case and hopefully for all similar cases".

Concerning the explanation, it looks like two threads are simultaneously in the scope where the pointer is not zero. Then one c deletes the pointer and the other one derefences null pointer.

Best regards,

Andrii

**Comment by andriish** [comment]
Here is the suspected code

```
static TPad * tmpGuideLinePad=nullptr;

   //delete all existing Guidelines and create new invisible pad
   if (tmpGuideLinePad) {
      if (object == tmpGuideLinePad) { // in case of funny button click combination.
         tmpGuideLinePad->Delete();
         tmpGuideLinePad = 0;
         return;
      }
      tmpGuideLinePad->Delete();
      tmpGuideLinePad = 0;
   }

   // Get Primitives
   prims = gPad->GetListOfPrimitives();
   n     = TMath::Min(15,prims->GetSize());
   Int_t lineColor = TColor::GetColor(239, 202, 0);

   TAttBBox2D *cur = dynamic_cast<TAttBBox2D *>( object );
   if (cur) {
      //create invisible TPad above gPad
      if (!tmpGuideLinePad){
         tmpGuideLinePad = new TPad("tmpGuideLinePad", "tmpGuideLinePad", 0, 0, 1, 1);
         Double_t x1, y1, x2, y2;
         gPad->GetRange(x1, y1, x2, y2);
         tmpGuideLinePad->Range(x1, y1, x2, y2);
         tmpGuideLinePad->SetFillStyle(0);
         tmpGuideLinePad->SetFillColor(0);
         tmpGuideLinePad->Draw();
         tmpGuideLinePad->cd();
         gPad->GetRange(x1, y1, x2, y2);
      }

```

**Comment by couet** [comment]
As I am not expert in that area I prefer to ask @pcanal approval.

**Comment by andriish** [comment]
Thanks!

**Comment by pcanal** [comment]
I am a bit lost ... Do we understand the reason for the thread unsafety?  What data is there a race condition on?  [Note for example that gPad is thread-local and in-itself not a cause for problem (unless several threads have their gPad pointing to the same pad).

**Comment by pcanal** [comment]
One thing that would be helpful to understand the scope is the 2 stack traces of the 2 threads accessing the same pointer.

File: graf2d/gpad/src/TPad.cxx [hunk 1]
```diff
@@ -6172,7 +6172,7 @@ void TPad::ShowGuidelines(TObject *object, const Int_t event, const char mode, c
    // is off we do show guide lines.
    if ((event == kArrowKeyRelease) || (event == kArrowKeyPress) ||
        !gEnv->GetValue("Canvas.ShowGuideLines", 0)) return;
-
+   R__LOCKGUARD(gROOTMutex);
```
Comment (pcanal) [inline graf2d/gpad/src/TPad.cxx:6175]:
```suggestion
-    static TPad * tmpGuideLinePad;
+   TPad * tmpGuideLinePad = dynamic_cast<TPad*>(gPad->GetListOfPrimitives()->FindObject("tmpGuideLinePad");
```

**Comment by guitargeek** [comment]
Quoting @pcanal:

> I am a bit lost ... Do we understand the reason for the thread unsafety? What data is there a race condition on? [Note for example that gPad is thread-local and in-itself not a cause for problem (unless several threads have their gPad pointing to the same pad).

> One thing that would be helpful to understand the scope is the 2 stack traces of the 2 threads accessing the same pointer.

Since you @andriish also found a workaround for your problem, it would be better to only change the code in ROOT once we understand exactly what the problem is and how it is solved by this change (or another change).

Therefore I'll close this PR here, but please feel free to engage with us again either here on GitHub or on the ROOT forum if you have practical problems again! Thanks

**Pull request closed without merging**